File: csharp/Pubnub.cs
using System;
using System.IO;
using System.Net;
using System.Text;
using System.Collections;
using NetServ.Net.Json;

public class PubnubTest {
    private static string channel       = "my_unique_channel";
    private static string PUBLISH_KEY   = "demo";
    private static string SUBSCRIBE_KEY = "demo";

    static public void Main() {
        // Example Publish Usage
        PubnubTest.test_publish1();
        PubnubTest.test_publish2();

        // Example History Usage
        PubnubTest.test_history();

        // Example Subscribe Usage
        PubnubTest.test_subscribe();
    }

    /// <summary>
    /// Test for Pubnub.publish( string, JsonObject )
    /// </summary>
    private static void test_publish1() {
        Console.WriteLine("TEST PUBLISH 1:");

        // Init Pubnub Class
        Pubnub pubnub  = new Pubnub(
            PubnubTest.PUBLISH_KEY,
            PubnubTest.SUBSCRIBE_KEY
        );

        JsonObject message = new JsonObject();

        // Add Key
        message.Add( "my_var", "Hello World 2!" );

        // Send (Publish) Message
        JsonObject response = pubnub.publish( PubnubTest.channel, message );

        // Print Stuff
        Console.WriteLine("Status: " + response["status"]);
        Console.WriteLine("-----------------------------------------------\n");
    }

    /// <summary>
    /// Test for Pubnub.publish( string, JsonObject )
    /// </summary>
    private static void test_publish2() {
        Console.WriteLine("TEST PUBLISH 2:");

        // Init Pubnub Class
        Pubnub pubnub  = new Pubnub(
            PubnubTest.PUBLISH_KEY,
            PubnubTest.SUBSCRIBE_KEY
        );

        JsonWriter writer  = new JsonWriter();
        JsonObject message = new JsonParser( new StringReader(
            "{\"my_var\" : \"Hello World!\"}"
        ), true ).ParseObject();

        // Send (Publish) Message
        JsonObject response = pubnub.publish( PubnubTest.channel, message );

        // Print Full Response
        response.Write(writer);
        Console.WriteLine("Publish() Response: " + writer.ToString());

        // Print Parts
        Console.WriteLine("Status: " + response["status"]);
        Console.WriteLine("-----------------------------------------------\n");
    }

    /// <summary>
    /// Test for Pubnub.history( string, int )
    /// </summary>
    private static void test_history() {
        Console.WriteLine("TEST HISTORY:");

        // Init Pubnub Class
        Pubnub pubnub  = new Pubnub(
            PubnubTest.PUBLISH_KEY,
            PubnubTest.SUBSCRIBE_KEY
        );

        // Get History
        JsonArray response = pubnub.history( PubnubTest.channel, 10 );
        JsonWriter writer  = new JsonWriter();

        // Print Response
        response.Write(writer);
        Console.WriteLine("History() Response: " + writer.ToString());

        // Print Each Message
        foreach (JsonObject message in response) {
            JsonWriter msg_writer = new JsonWriter();
            message.Write(msg_writer);
            Console.WriteLine(msg_writer.ToString());
        }
        Console.WriteLine("-----------------------------------------------\n");
    }

    /// <summary>
    /// Test for Pubnub.subscribe( string, delegate )
    /// </summary>
    private static void test_subscribe() {
        Console.WriteLine("TEST SUBSCRIBE:");
        Console.WriteLine("TO TEST THIS FUNCTION, PUBLISH MESSAGES:");

        // Init Pubnub Class
        Pubnub pubnub  = new Pubnub(
            PubnubTest.PUBLISH_KEY,
            PubnubTest.SUBSCRIBE_KEY
        );

        // Subscribe
        pubnub.subscribe(
            PubnubTest.channel,
            delegate (JsonObject message) {
                // Print Message
                JsonWriter writer = new JsonWriter();
                message.Write(writer);
                Console.WriteLine("Subscribe() Response: "+writer.ToString());

                // Continue Listening
                return true;
            }
        );
        Console.WriteLine("-----------------------------------------------\n");
    }
}


public class Pubnub {
    private static string ORIGIN        = "http://pubnub-prod.appspot.com";
    private static int    LIMIT         = 1700;
    private static int    UNIQUE        = 1;
    private static string PUBLISH_KEY   = "";
    private static string SUBSCRIBE_KEY = "";

    public delegate bool Procedure(JsonObject message);

    public Pubnub( string publish_key, string subscribe_key ) {
        Pubnub.PUBLISH_KEY   = publish_key;
        Pubnub.SUBSCRIBE_KEY = subscribe_key;
    }

    public JsonArray history( string channel, int limit ) {
        channel = Pubnub.SUBSCRIBE_KEY + "/" + channel;
        Hashtable hist_params = new Hashtable();

        hist_params.Add( "channel", channel );
        hist_params.Add( "limit", limit.ToString() );

        // Get History
        JsonObject response = this._request(
            Pubnub.ORIGIN + "/pubnub-history",
            hist_params
        );

        // Return Array of Historical Messages
        return (JsonArray)response["messages"];
    }

    public JsonObject publish( string channel, JsonObject message ) {
        channel = Pubnub.SUBSCRIBE_KEY + "/" + channel;
        Hashtable pub_params = new Hashtable();
        JsonWriter writer    = new JsonWriter();

        message.Write(writer);

        pub_params.Add( "publish_key", Pubnub.PUBLISH_KEY );
        pub_params.Add( "channel", channel );
        pub_params.Add( "message", writer.ToString() );

        // Publish Message Message
        return this._request( Pubnub.ORIGIN + "/pubnub-publish", pub_params );
    }

    private void _subscribe(
        string channel,
        Procedure callback,
        string timetoken,
        string server
    ) {
        bool keep_listening = true;

        // Find a PubNub Server
        if (server.Length == 0) {
            Hashtable sub_params = new Hashtable();

            sub_params.Add( "channel", channel );

            JsonObject resp_for_server = this._request(
                Pubnub.ORIGIN + "/pubnub-subscribe",
                sub_params
            );

            try {
                server = resp_for_server["server"].ToString();
            }
            catch {
                return;
            }
        }

        try {
            Hashtable conn_params = new Hashtable();

            conn_params.Add( "channel", channel );
            conn_params.Add( "timetoken", timetoken );

            // Listen for a Message
            JsonObject response = this._request(
                "http://" + server + "/",
                conn_params
            );

            // Test for Message
            if (((JsonArray)response["messages"]).Count == 0) {
                this._subscribe( channel, callback, timetoken, "" );
                return;
            }

            // Was it a Timeout
            JsonArray messages = (JsonArray)response["messages"];
            if (messages[0].ToString() == "xdr.timeout") {
                timetoken = response["timetoken"].ToString();
                this._subscribe( channel, callback, timetoken, server );
                return;
            }

            // Run user Callback and Reconnect if user permits.
            foreach (JsonObject message in messages) {
                keep_listening = keep_listening && callback(message);
            }

            // Keep listening if Okay.
            if (keep_listening) {
                timetoken = response["timetoken"].ToString();
                this._subscribe( channel, callback, timetoken, server );
                return;
            }
        }
        catch {
            this._subscribe( channel, callback, timetoken, "" );
            return;
        }

        // Done Listening.
        return;
    }

    public void subscribe( string channel, Procedure callback ) {
        channel = Pubnub.SUBSCRIBE_KEY + "/" + channel;
        this._subscribe( channel, callback, "0", "" );
    }

    private JsonObject _request( string url, Hashtable args ) {
        StringBuilder  sb = new StringBuilder();
        byte[]        buf = new byte[8192];

        // Add Unique
        args.Add( "unique", (Pubnub.UNIQUE).ToString() );

        // Build URL
        url += "?";
        foreach (DictionaryEntry dict in args) {
            url += System.Uri.EscapeDataString((string)dict.Key) + "=" +
                   System.Uri.EscapeDataString((string)dict.Value) + "&";
        }
        url = url.Substring( 0, url.Length -1 );

        if (url.Length > Pubnub.LIMIT) {
            return new JsonParser( new StringReader(
                "{\"message\" : \"Message Too Long\"}"
            ), true ).ParseObject();
        }

        // Create Request
        HttpWebRequest  request   = (HttpWebRequest)WebRequest.Create(url);
        HttpWebResponse response  = (HttpWebResponse)request.GetResponse();
        Stream          resStream = response.GetResponseStream();

        string tempString = null;
        int    count      = 0;

        do {
            count = resStream.Read( buf, 0, buf.Length );
            if (count != 0) {
                tempString = Encoding.ASCII.GetString( buf, 0, count );
                sb.Append(tempString);
            }
        } while (count > 0);

        // Parse Response
        string message = sb.ToString();
               message = message.Substring( 10, message.Length - 11 );

        JsonParser parser = new JsonParser( new StringReader(message), true );
        JsonObject json   = parser.ParseObject();

        // Return Response
        return json;
    }
}
 

File: java/pubnub/Pubnub.java
package pubnub;
import org.json.*;
import java.util.HashMap;
import java.util.Iterator;
import java.net.URL;
import java.net.URLEncoder;
import java.io.BufferedReader;
import java.io.InputStreamReader;

public class Pubnub {
    private static String ORIGIN        = "http://pubnub-prod.appspot.com";
    private static int    LIMIT         = 1700;
    private static int    UNIQUE        = 1;
    private static String PUBLISH_KEY   = "";
    private static String SUBSCRIBE_KEY = "";

    public Pubnub( String publish_key, String subscribe_key ) {
        Pubnub.PUBLISH_KEY   = publish_key;
        Pubnub.SUBSCRIBE_KEY = subscribe_key;
    }

    public JSONObject publish( String channel, JSONObject message ) {
        HashMap<String,String> params = new HashMap<String,String>();
        channel = Pubnub.SUBSCRIBE_KEY + "/" + channel;

        // Prepare HTTP GET Params
        params.put( "publish_key", Pubnub.PUBLISH_KEY );
        params.put( "channel", channel );
        params.put( "message", message.toString() );

        return this._request( Pubnub.ORIGIN + "/pubnub-publish", params );
    }

    public void subscribe( String channel, Callback callback ) {
        channel = Pubnub.SUBSCRIBE_KEY + "/" + channel;
        this._subscribe( channel, callback, "0", "" );
    }

    private void _subscribe(
        String   channel,
        Callback callback,
        String   timetoken,
        String   server
    ) {
        boolean keep_listening = true;

        // Find a PubNub Server
        if (server.length() == 0) {
            HashMap<String,String> sub_params = new HashMap<String,String>();

            sub_params.put( "channel", channel );

            JSONObject resp_for_server = this._request(
                Pubnub.ORIGIN + "/pubnub-subscribe",
                sub_params
            );

            server = resp_for_server.optString("server");
        }

        try {
            HashMap<String,String> conn_params = new HashMap<String,String>();

            // Prepare HTTP GET Params
            conn_params.put( "channel", channel );
            conn_params.put( "timetoken", timetoken );

            // Listen for a Message
            JSONObject response = this._request(
                "http://" + server + "/",
                conn_params
            );

            // Capture Messages
            JSONArray messages = response.optJSONArray("messages");

            // Test for Message
            if (messages.length() == 0) {
                this._subscribe( channel, callback, timetoken, "" );
                return;
            }

            // Was it a Timeout
            if (messages.optString( 0, "n" ).equals("xdr.timeout")) {
                timetoken = response.optString("timetoken");
                this._subscribe( channel, callback, timetoken, server );
                return;
            }

            // Run user Callback and Reconnect if user permits.
            for ( int i = 0; messages.length() > i; i++ ) {
                JSONObject message = messages.optJSONObject(i);
                keep_listening = keep_listening && callback.execute(message);
            }

            // Keep listening if Okay.
            if (keep_listening) {
                timetoken = response.optString("timetoken");
                this._subscribe( channel, callback, timetoken, server );
                return;
            }
        }
        catch (Exception e) {
            this._subscribe( channel, callback, timetoken, "" );
            return;
        }
    }

    public JSONArray history( String channel, int limit ) {
        HashMap<String,String> params = new HashMap<String,String>();
        channel = Pubnub.SUBSCRIBE_KEY + "/" + channel;

        // Prepare HTTP GET Params
        params.put( "channel", channel );
        params.put( "limit", Integer.toString(limit) );

        JSONObject response = this._request(
            Pubnub.ORIGIN + "/pubnub-history",
            params
        );

        return response.optJSONArray("messages");
    }

    public JSONObject _request( String url, HashMap<String,String> args ) {
        // Add Unique Param
        args.put( "unique", Integer.toString(Pubnub.UNIQUE) );

        String   json         = "";
        Iterator arg_iterator = args.keySet().iterator();

        url += "?";
        while (arg_iterator.hasNext()) {
            try {
                String key = (String)arg_iterator.next();
                url += URLEncoder.encode( key, "UTF-8" ) + "=" +
                       URLEncoder.encode( args.get(key), "UTF-8" ) + "&";
            }
            catch(java.io.UnsupportedEncodingException e) {
                e.printStackTrace();
                JSONObject jsono = new JSONObject();
                try { jsono.put( "message", "Failed UTF-8 Encoding URL." ); }
                catch (org.json.JSONException jsone) {}
                return jsono;
            }
        }
        url = url.substring( 0, url.length() -1 );

        // Fail if string too long
        if (url.length() > Pubnub.LIMIT) {
            JSONObject jsono = new JSONObject();
                try { jsono.put( "message", "Message Too Long." ); }
                catch (org.json.JSONException jsone) {}
            return jsono;
        }

        try {
            URL            request = new URL(url);
            String         line    = "";
            String         jsonp   = "";
            BufferedReader reader  = new BufferedReader(
                new InputStreamReader(request.openStream())
            );

            // Read JSONP Message
            while ((line = reader.readLine()) != null) { jsonp += line; }
            reader.close();

            // Remove JSONP Wrapper
            json = jsonp.substring( 10, jsonp.length() - 1 );
            // System.out.println(jsonp);

        } catch (Exception e) {
            JSONObject jsono = new JSONObject();

            try { jsono.put( "message", "Failed JSONP HTTP Request." ); }
            catch (org.json.JSONException jsone) {}

            e.printStackTrace();
            System.out.println(e);

            return jsono;
        }

        // System.out.println(json);

        // Parse JSON String
        try { return new JSONObject(json); }
        catch (org.json.JSONException jsone) {}

        return new JSONObject();
    }
}

